flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskFailureHistoryEntry.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.executiongraph.ErrorInfo;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+
+/**
+ * {@code TaskFailureHistoryEntry} collects information about a single task failure that should be
+ * exposed through the exception history.
+ */
+public class TaskFailureHistoryEntry implements Serializable {
+
+    private static final long serialVersionUID = -3855285510064263701L;
+
+    private final ErrorInfo errorInfo;
+    @Nullable private final String failingTaskName;
+    @Nullable private final TaskManagerLocation assignedResourceLocation;
+
+    private TaskFailureHistoryEntry(
+            ErrorInfo errorInfo,
+            @Nullable String failingTaskName,
+            @Nullable TaskManagerLocation assignedResourceLocation) {
+        this.errorInfo = Preconditions.checkNotNull(errorInfo);
+        this.failingTaskName = failingTaskName;
+        this.assignedResourceLocation = assignedResourceLocation;
+    }
+
+    public boolean isGlobalFailure() {
+        return failingTaskName != null;
+    }
+
+    public ErrorInfo getErrorInfo() {
+        return errorInfo;
+    }
+
+    @Nullable
+    public String getFailingTaskName() {
+        return failingTaskName;
+    }
+
+    @Nullable
+    public TaskManagerLocation getAssignedResourceLocation() {
+        return assignedResourceLocation;
+    }
+
+    public static TaskFailureHistoryEntry createGlobalFailureEntry(
+            Throwable failureCause, long timestamp) {
+        return new TaskFailureHistoryEntry(new ErrorInfo(failureCause, timestamp), null, null);
```

Comment:
Seems like `TaskFailureHistoryEntry` is not an appropriate name? Why not just extend the `ErrorInfo` instead?

Reply:
I renamed it `ExceptionHistoryEntry`. I hesitated to merge `ErrorInfo` and `ExceptionHistoryEntry` because I wanted to have the exception history decoupled from the `Execution`? The `Execution` only needs to know about the failure and the timestamp.